modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java

```diff
@@ -154,6 +155,33 @@ public boolean needs_score() {
                 }
             };
             return context.factoryClazz.cast(factory);
+        } else if (context.instanceClazz.equals(ScoreScript.class)) {
```

Comment:
I'm confused on why this case is added. I see a couple issues with it:
* GenericElasticsearchScript is about to be removed (see https://github.com/elastic/elasticsearch/pull/34730), and all the uses of it (through SearchScript) have already been removed
* We already have caching of the compiled script in ScriptService, so why do we need special handling of caching here?